src/java.desktop/share/classes/javax/swing/text/DefaultCaret.java

```diff
@@ -379,8 +380,18 @@ protected void moveCaret(MouseEvent e) {
     public void focusGained(FocusEvent e) {
         if (component.isEnabled()) {
             if (component.isEditable()) {
-                setVisible(true);
+                if (savedBlinkRate != 0) {
+                    setBlinkRate(savedBlinkRate);
+                    savedBlinkRate = 0;
+                }
```

Comment:
My general questions need to be addressed first, but after we get past that you need to revisit this code.
It seems to overlook that an application could change the blink rate itself, and restores the saved rate over what
the application has set.

Also, SFAICT setBlinkRate() doesn't seem to be preventing negative values ...

Reply:
> My general questions need to be addressed first, but after we get past that you need to revisit this code. It seems to overlook that an application could change the blink rate itself, and restores the saved rate over what the application has set.

Ok, the only way it can be a problem is when user will try to set the blink rate of the uneditable component's caret. I think i will have to special-case that.

> Also, SFAICT setBlinkRate() doesn't seem to be preventing negative values ...

I wonder what will happen if i set a negative rate. Will it blink into the past before component is even created? Can it cause a time paradox? Hmm.